dask/array/optimization.py

```diff
@@ -0,0 +1,166 @@
+    ...        'b': (getitem, 'a', (slice(None, None, None),)),
+    ...        'c': (getitem, 'b', (slice(None, None, None),)),
+    ...        'd': (getitem, 'c', (slice(None, 5, None),)),
+    ...        'e': (getitem, 'd', (slice(None, None, None),))}
+
+    >>> remove_full_slices(dsk)  # doctest: +SKIP
+    {'a': (range, 5),
+     'e': (getitem, 'a', (slice(None, 5, None),))}
+    """
+    full_slice_keys = set(k for k, task in dsk.items() if is_full_slice(task))
+    dsk2 = dict((k, task[1] if k in full_slice_keys else task)
+                 for k, task in dsk.items())
+    dsk3 = dealias(dsk2)
+    return dsk3
+
+
+a, b, x = '~a', '~b', '~x'
+
+
+def fuse_slice_dict(match):
+    # Making new dimensions?  Need two getitems
+    # TODO: well, we could still optimize the non-None axes
+
+    try:
+        c = fuse_slice(match[a], match[b])
+        return (getitem, match[x], c)
+    except NotImplementedError:
+        return (getitem, (getitem, match[x], match[a]),
+                         match[b])
+
+
+rewrite_rules = RuleSet(RewriteRule((getitem, (getitem, x, a), b),
+                                    fuse_slice_dict,
+                                    (a, b, x)))
+
+
+def normalize_slice(s):
+    start, stop, step = s.start, s.stop, s.step
+    if start is None:
+        start = 0
+    if step is None:
+        step = 1
+    return slice(start, stop, step)
+
+
+def fuse_slice(a, b):
+    """ Fuse stacked slices together
+
+    >>> fuse_slice(slice(1000, 2000), slice(10, 15))
+    slice(1010, 1015, None)
+
+    >>> fuse_slice(slice(1000, 2000), 10)
+    1010
+
+    >>> fuse_slice(slice(1000, 2000, 5), 10)
+    1050
+
+    >>> fuse_slice(slice(1000, 2000, 5), slice(10, 20, 2))
+    slice(1050, 1100, 10)
+
+    >>> fuse_slice(None, slice(None, None))
+    None
+    """
+    if a is None and b == slice(None, None):
+        return None
+    if isinstance(a, slice):
+        a = normalize_slice(a)
+    if isinstance(b, slice):
+        b = normalize_slice(b)
+
+    if isinstance(a, slice) and isinstance(b, int):
+        return a.start + b*(a.step if a.step is not None else 1)
+    if isinstance(a, slice) and isinstance(b, slice):
+        start = a.start + a.step * b.start
+        if b.stop is not None:
+            stop = a.start + a.step * b.stop
+        else:
+            stop = None
+        if a.stop is not None:
+            if stop is not None:
+                stop = min(a.stop, stop)
+            else:
+                stop = a.stop
+            stop = stop
+        step = a.step * b.step
+        if step == 1:
+            step = None
+        return slice(start, stop, step)
+    if isinstance(a, tuple) and not isinstance(b, tuple):
+        b = (b,)
+    if isinstance(a, tuple) and isinstance(b, tuple):
+        j = 0
+        result = list()
+        for i in range(len(a)):
+            if isinstance(a[i], int):
+                result.append(a[i])
+                continue
+            if (isinstance(a, slice) and
```

Comment:
Refresh, that has been moved out.

On Sat, Mar 28, 2015 at 1:53 PM, Stephan Hoyer <notifications@github.com>
wrote:

> In dask/array/optimization.py
> <https://github.com/ContinuumIO/dask/pull/107#discussion_r27348387>:
>
> > +                stop = a.stop
> > +            stop = stop
> > +        step = a.step * b.step
> > +        if step == 1:
> > +            step = None
> > +        return slice(start, stop, step)
> > +    if isinstance(a, tuple) and not isinstance(b, tuple):
> > +        b = (b,)
> > +    if isinstance(a, tuple) and isinstance(b, tuple):
> > +        j = 0
> > +        result = list()
> > +        for i in range(len(a)):
> > +            if isinstance(a[i], int):
> > +                result.append(a[i])
> > +                continue
> > +            if (isinstance(a, slice) and
>
> how can you even get here? you just checked isinstance(a, tuple) above.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ContinuumIO/dask/pull/107/files#r27348387>.
>